crates/uv-auth/src/middleware.rs

```diff
@@ -100,88 +128,184 @@ impl Middleware for AuthMiddleware {
         };
         trace!("Handling request for {url}");
 
-        // Then check for credentials in (2) the cache
-        let credentials = self.cache().check(request.url(), credentials);
-
-        // Track credentials that we might want to insert into the cache
-        let mut new_credentials = None;
-
-        // If already authenticated (including a password), don't query other services
-        if credentials.is_authenticated() {
-            match credentials {
-                // If we get credentials from the cache, update the request
-                CheckResponse::Cached(credentials) => request = credentials.authenticate(request),
-                // If we get credentials from the request, we should update the cache
-                // but don't need to update the request
-                CheckResponse::Uncached(credentials) => new_credentials = Some(credentials.clone()),
-                CheckResponse::None => unreachable!("No credentials cannot be authenticated"),
+        if let Some(credentials) = credentials {
+            let credentials = Arc::new(credentials);
+
+            // If there's a password, send the request and cache
+            if credentials.password().is_some() {
+                trace!("Request for {url} is already fully authenticated");
+                self.complete_request(Some(credentials), request, extensions, next)
+                    .await
+            } else {
+                trace!("Request for {url} is missing a password, looking for credentials");
+                // There's just a username, try to find a password
+                let credentials = if let Some(credentials) = self
+                    .cache()
+                    .get_realm(NetLoc::from(request.url()), credentials.to_username())
```

Comment:
So in this path, we can use the realm because we know we need credentials?

Reply:
Yeah we know that authentication is needed so we don't need to "try" the request first. We could check for cached URL-level credentials first, that seems more "correct" but I feel like it'd be surprising to have the same username for multiple URLs in the same realm.